message_feed: Added `Quote and reply or forward`  button in the message_header_bar.

palashb01 opened the pull request:
*Added a "Quote and reply or forward" button (with the appropriate tooltip) to the message header bar, just to the left of the three-dot menu.
*Removed "Quote and reply or forward" from the three-dot message menu.

<!-- Describe your pull request here.-->

Fixes: #23100 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

![Screenshot_20221002_002119](https://user-images.githubusercontent.com/66828942/193424089-cbc47828-da85-4213-bd6b-3431793a1c40.png)


![Screenshot_20221002_002219](https://user-images.githubusercontent.com/66828942/193424102-c8e23440-2d83-46b7-81e1-0fb21343671b.png)


**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

palashb01 commented:
Hey @alya @timabbott @sahil839 , Please have a look at it, Thanks.

sahil839 commented on static/js/click_handlers.js:
```diff
@@ -280,6 +280,17 @@ export function initialize() {
+        // compose_actions.respond_to_message doesn't take a message
+        // argument).
+        compose_actions.quote_and_reply({trigger: "popover respond"});
+        // hide_actions_popover();
```
Why is this commented?

sahil839 commented on static/templates/message_controls.hbs:
```diff
@@ -9,6 +9,12 @@
     </div>
     {{/unless}}
 
+    {{#unless should_display_quote_and_reply}}
```
You are still not passing `should_display_quote_and_reply` to this template and it will always be `undefined`,

palashb01 replied:
Could you Please elaborate, how `should_display_quote_and_reply` would be undefined, Thanks.

sahil839 replied:
See `should_display_quote_and_reply` is included in the args which is passed to `render_actions_popover_content` (for `actions_popover_content.hbs` file) [here](https://github.com/zulip/zulip/blob/da8817889bbd024656b39db0c98f30b95bd4de82/static/js/popovers.js#L554). Similarly you have to pass it in the `message_controls` template as well, otherwise the template does not know its value and considers it as undefined.

sahil839 reviewed (commented):
Posted a couple of comments. I can also see a zulip submodule in `Files changed`, not sure why.

Also, please fix the CI failure.

palashb01 commented:
Hey @sahil839 @N-Shar-ma , Please have a look at it, Thanks.

sahil839 reviewed (commented):
Left some comments.

Also please fix the CI to pass and squash the commits as it should ideally be one commit only. I also noticed you are working on `main` branch , it would be better if you work on a separate feature branch for each issue. And please use `git fetch` and then `git rebase` and not `git pull` for updating your fork, since we do not use merge commits. 

You can refer to the following docs - https://zulip.readthedocs.io/en/latest/git/using.html# and https://zulip.readthedocs.io/en/latest/git/pull-requests.html#id1

palashb01 commented on static/js/message_list_view.js:
```diff
@@ -756,6 +757,10 @@ export class MessageListView {
 
             message.url = hash_util.by_conversation_and_time_url(message);
 
+            message.should_display_quote_and_reply=(()=>{
```
Okay,got it Thanks a lot.

palashb01 commented on static/js/popovers.js:
```diff
@@ -1227,18 +1223,6 @@ export function register_click_handlers() {
-        hide_actions_popover();
-        e.stopPropagation();
-        e.preventDefault();
-    });
```
**It is working locally the button and the shortcut both, I have attached the video of the following, Should I add the click handler ?**

https://user-images.githubusercontent.com/66828942/195880603-75c41a76-b394-4aee-a5f5-b9637cbb648a.mp4

sahil839 replied:
I think you have removed the click handler again while updating the PR. Neither I can see the code which will do something on clicking the `Quote and reply` button you added nor it works locally.

sahil839 commented:
Please fix the commit history first as I asked above, it is becoming difficult to review and test this locally on my system.

palashb01 commented:
Hey @sahil839 !  Please Review my PR, Thanks.

sahil839 commented on static/styles/zulip.css:
```diff
@@ -1399,7 +1400,12 @@ td.pointer {
 }
 
-.actions_hover:hover {
+.popover_info i.zulip-icon.zulip-icon-bot {
```
But where is this element, `zulip-icon-bot`, I cannot see this class in any of the four icons. Plus there is no `popover_info` element currently.

sahil839 reviewed (commented):
Posted a couple of comments. Also please update the commit message to mention what is this commit for actually.

palashb01 commented:
Hey @sahil839 , Please have a look at it, also I have posted some doubts on  [CZO](https://chat.zulip.org/#narrow/pm-with/10242-user10242) Please consider them too, Thanks.

palashb01 commented:
Hey @sahil839, I have resolved the discussed issue on CZO, and have resolved the issue assigned to me.

- [x] Added a "Quote and reply or forward" button (with the appropriate tooltip) to the message header bar, just to the left of the three-dot menu.
- [x] Removed "Quote and reply or forward" from the three-dot message menu.

**Demo:**

https://user-images.githubusercontent.com/66828942/198042571-cc45881c-f23e-45a2-a488-90b071e17676.mp4

Please review the PR, Thanks.

alya commented:
@palashb01 could you please post a video demoing what happens at different screen sizes, where we might currently have the buttons very close to the message text?

palashb01 commented:
Hey @alya , here are the demo videos you asked for:

**Current CSS with the buttons:**


https://user-images.githubusercontent.com/66828942/198142029-f3300ad7-4d0a-4dcf-be6e-cedec0917f83.mp4




**New changes with the buttons:**


https://user-images.githubusercontent.com/66828942/198141370-680572f9-e4d0-45b6-8f4e-f5b169dcf38c.mp4




**-The current CSS places the `emoji` and `quote and reply or forward` icons too close together, that reduces visibility in mobile screen size.**

**- With the updated CSS, icons now have sufficient space between them, improving readability and usability.**

alya commented:
Taking a careful look at this, the icons are spaced much more closely than they are at present, which makes them hard to use. We should keep the spacing as it currently is, or close to it.

palashb01 commented:
Hey @alya  , I made some adjustments based on your product review so that there is now more space between the icons and they more closely match the existing icons; I've attached a screenshot for your assessment; I'll commit the modifications once you've looked them over, Thanks.

**ScreenShot:**

![zulip_23143](https://user-images.githubusercontent.com/66828942/199166258-b67d72e9-5d95-442a-bc72-e10794c16141.png)

![zulip_23143_2](https://user-images.githubusercontent.com/66828942/199166264-e81b7c0f-8c3c-4aab-b9a1-bea4743a7ef3.png)

alya commented:
Thanks! That certainly looks better, but can you clarify what you mean by "more closely match"? Is the spacing the same as now, or not? If not, how did you select it?

palashb01 commented:
> Thanks! That certainly looks better, but can you clarify what you mean by "more closely match"? Is the spacing the same as now, or not? If not, how did you select it?

Yes, the spacing is same as it is now, I will push the changes.

palashb01 commented:
Hey @alya , I have made the necessary changes, Please review it Thanks.

**Screenshot:**

![zulip_23143_2](https://user-images.githubusercontent.com/66828942/199166264-e81b7c0f-8c3c-4aab-b9a1-bea4743a7ef3.png)

palashb01 commented:
Hey @alya @sahil839 , could you Please review the PR and suggest, if some changes are needed, Thanks.

sahil839 commented:
The icons are closer to the timestamp than before. I think we can keep some space between them as it was before.

palashb01 commented:
> The icons are closer to the timestamp than before. I think we can keep some space between them as it was before.

Sure, I will make the desired changes and commit the same, Thanks.

sahil839 commented:
UI looks good, but I am not sure of the CSS changes (since we use hardcoded values). Probably would be better to do it after #23456 is done.

palashb01 commented:
Closing this PR to open a new one, since this PR was raised against the main branch. A new one will be opened with a dedicated branch. Also, the CSS codebase and design have changed a lot over the past few days, it would be better to start with a new PR.
